src/criticality/BoratedWater.C

```diff
@@ -0,0 +1,223 @@
+  }
+
+  if (!full_names.empty())
+  {
+    std::ostringstream msg;
+    msg << "The criticality search will clear out all nuclides in material "
+        << std::to_string(_material_id)
+        << " and replace them with the naturally-abundant nuclides in hydrogen, oxygen, and boron. "
+           "Any other nuclides which existed in the material will be deleted."
+        << std::endl;
+    msg << "\nThe material you provided contains nuclides which are not the natural isotopes of H, "
+           "B, and O: "
+        << full_names.substr(0, full_names.length() - 1) << ". These will be deleted from material "
+        << std::to_string(_material_id) << " when the boron concentration is changed." << std::endl;
+    msg << "\nFor general criticality searches based on material composition, please contact the "
+           "Cardinal developer team.";
+    mooseWarning(msg.str());
+  }
+
+  // Compute the molar mass of pure water
+  _M_H2O = 0.0;
+  for (const auto & h : _hydrogen_natural)
+    _M_H2O += 2.0 * h.second * NuclearData::Nuclide::getAtomicMass(h.first);
+  for (const auto & o : _oxygen_natural)
+    _M_H2O += 1.0 * o.second * NuclearData::Nuclide::getAtomicMass(o.first);
+
+  // Compute the molar mass of pure boron
+  _M_B = 0.0;
+  for (const auto & b : _boron_natural)
+    _M_B += 1.0 * b.second * NuclearData::Nuclide::getAtomicMass(b.first);
+}
+
+void
+BoratedWater::updateOpenMCModel(const Real & ppm)
+{
+  _console << "Searching for boron = " << ppm << " [ppm] ..." << std::endl;
+
+  // A coupled thermal-fluid app may set the material density just before we enter
+  // this routine; we will preserve that density which may be set and we interpret
+  // it to be the SOLUTION density to be fully consistent with energy conservation.
+
+  // Compute the number fractions of each element
+  Real frac_H2O = (1 - ppm * 1e-6) / _M_H2O;
+  Real frac_H = 2 * frac_H2O;
+  Real frac_O = frac_H2O;
+  Real frac_B = ppm * 1e-6 / _M_B;
+
+  double rho;
+  int err = openmc_material_get_density(_material_index, &rho);
+  catchOpenMCError(err, "get density for material " + std::to_string(_material_id));
+
+  std::vector<std::string> names;
+  std::vector<double> densities;
+  for (const auto & h : _hydrogen_natural)
+  {
+    names.push_back(h.first);
+    densities.push_back(h.second * frac_H * rho * openmc::N_AVOGADRO);
+  }
+  for (const auto & o : _oxygen_natural)
+  {
+    names.push_back(o.first);
+    densities.push_back(o.second * frac_O * rho * openmc::N_AVOGADRO);
+  }
+  for (const auto & b : _boron_natural)
+  {
+    names.push_back(b.first);
+    densities.push_back(b.second * frac_B * rho * openmc::N_AVOGADRO);
+  }
+
+  openmc::model::materials[_material_index]->set_densities(names, densities /* atom/b-cm */);
+
+  const int * nucl;
+  const double * dens;
+  int n;
+  err = openmc_material_get_densities(_material_index, &nucl, &dens, &n);
+  catchOpenMCError(err, "get nuclide densities from material " + std::to_string(_material_id));
```

Comment:
Is this just checking that we haven't left the material in some broken state?

Reply:
I guess my question was larger than that... there is no algorithmic reason to call `openmc_material_get_densities()` here.  We have already set them, and don't do anything with these results.  I therefore assumed that we called this to just test that we set things properly?